File: 240421/강력한 폭발/strong-explosion.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    static int N;
    static int[][] grid;
    static int result = 0, bombCnt;
    static List<int[]> bombList = new ArrayList<>();

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();
        grid = new int[N][N];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                int tmp = grid[i][j] = scanner.nextInt();
                if (tmp == 1) {
                    bombList.add(new int[] {i, j});
                }
            }
        }
        bombCnt = bombList.size();
        computeCombinations(0, new int[bombCnt]);
        System.out.println(result);
    }

    static boolean inRange(int r, int c) {
        return r >= 0 && r < N && c >= 0 && c < N;
    }

    static int applyATypeBomb(int r, int c, int[][] copyGrid) {
        int count = 0;
        int[] rows = {-2, -1, 1, 2};
        for (int delta : rows) {
            int nr = r + delta;
            if (inRange(nr, c) && copyGrid[nr][c] == 0) {
                copyGrid[nr][c] = 1;
                count++;
            }
        }
        return count;
    }

    static int applyBTypeBomb(int r, int c, int[][] copyGrid) {
        int count = 0;
        int[][] directions = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        for (int[] dir : directions) {
            int nr = r + dir[0];
            int nc = c + dir[1];
            if (inRange(nr, nc) && copyGrid[nr][nc] == 0) {
                copyGrid[nr][nc] = 1;
                count++;
            }
        }
        return count;
    }

    static int applyCTypeBomb(int r, int c, int[][] copyGrid) {
        int count = 0;
        int[][] directions = {{-1, -1}, {-1, 1}, {1, 1}, {1, -1}};
        for (int[] dir : directions) {
            int nr = r + dir[0];
            int nc = c + dir[1];
            if (inRange(nr, nc) && copyGrid[nr][nc] == 0) {
                copyGrid[nr][nc] = 1;
                count++;
            }
        }
        return count;
    }

    static void computeCombinations(int depth, int[] selected) {
        if (depth == bombCnt) {

            int[][] copyGrid = new int[N][N];
            for (int i = 0; i < N; i++) {
                System.arraycopy(grid[i], 0, copyGrid[i], 0, N);
            }

            int maxBombCount = bombCnt;
            for (int i = 0; i < bombCnt; i++) {
                int[] bomb = bombList.get(i);
                int bombType = selected[i];
                switch (bombType) {
                    case 1:
                        maxBombCount += applyATypeBomb(bomb[0], bomb[1], copyGrid);
                        break;
                    case 2:
                        maxBombCount += applyBTypeBomb(bomb[0], bomb[1], copyGrid);
                        break;
                    case 3:
                        maxBombCount += applyCTypeBomb(bomb[0], bomb[1], copyGrid);
                        break;
                }
            }
            result = Math.max(result, maxBombCount);
            return;
        }

        for (int i = 1; i <= 3; i++) {
            selected[depth] = i;
            computeCombinations(depth + 1, selected);
        }
    }
}

File: 240421/강력한 폭발/strong-explosion.py
N = int(input())
grid = [
    list(map(int, input().split())) 
    for _ in range(N)
    ]

bomb_list = [(r, c) for r in range(N) for c in range(N) if grid[r][c] == 1]

def in_range(r, c):
    return 0 <= r < N and 0 <= c < N

def a_type_bomb(r, c, copy_grid):
    cnt = 0
    # 위아래 2칸씩
    if in_range(r - 2, c) and copy_grid[r - 2][c] == 0:
        copy_grid[r - 2][c] = 1
        cnt += 1
    if in_range(r - 1, c) and copy_grid[r - 1][c] == 0:
        copy_grid[r - 1][c] = 1
        cnt += 1
    if in_range(r + 1, c) and copy_grid[r + 1][c] == 0:
        copy_grid[r + 1][c] = 1
        cnt += 1
    if in_range(r + 2, c) and copy_grid[r + 2][c] == 0:
        copy_grid[r + 2][c] = 1
        cnt += 1
    
    return cnt

def b_type_bomb(r, c, copy_grid):
    cnt = 0
    # 상하좌우
    if in_range(r - 1, c) and copy_grid[r - 1][c] == 0:
        copy_grid[r - 1][c] = 1
        cnt += 1
    if in_range(r + 1, c) and copy_grid[r + 1][c] == 0:
        copy_grid[r + 1][c] = 1
        cnt += 1
    if in_range(r, c + 1) and copy_grid[r][c + 1] == 0:
        copy_grid[r][c + 1] = 1
        cnt += 1
    if in_range(r, c - 1) and copy_grid[r][c - 1] == 0:
        copy_grid[r][c - 1] = 1
        cnt += 1

    return cnt


def c_type_bomb(r, c, copy_grid):
    cnt = 0
    # 좌상우상좌하우하
    if in_range(r - 1, c - 1) and copy_grid[r - 1][c - 1] == 0:
        copy_grid[r - 1][c - 1] = 1
        cnt += 1
    if in_range(r - 1, c + 1) and copy_grid[r - 1][c + 1] == 0:
        copy_grid[r - 1][c + 1] = 1
        cnt += 1
    if in_range(r + 1, c + 1) and copy_grid[r + 1][c + 1] == 0:
        copy_grid[r + 1][c + 1] = 1
        cnt += 1
    if in_range(r + 1, c - 1) and copy_grid[r + 1][c - 1] == 0:
        copy_grid[r + 1][c - 1] = 1
        cnt += 1

    return cnt

res = 0
selected = []

def comb(depth):
    if (depth == len(bomb_list)):
        global res
        copy_grid = [row[:] for row in grid]
        max_bomb_cnt = len(bomb_list)
        for (r, c), bomb_type in zip(bomb_list, selected):
            if bomb_type == 1:
                max_bomb_cnt += a_type_bomb(r, c, copy_grid)  
            elif bomb_type == 2:
                max_bomb_cnt += b_type_bomb(r, c, copy_grid)
            elif bomb_type == 3:
                max_bomb_cnt += c_type_bomb(r, c, copy_grid)
        res = max(res, max_bomb_cnt)
        return
    
    for i in range(1, 4):
        selected.append(i)
        comb(depth + 1)
        selected.pop()

comb(0)
print(res)